### DISCUSSION TITLE
Using a shift-in register

### DISCUSSION THREAD

**Discussion by reznet** [tttapa/Control-Surface#389] [Q&A]
Hi, I'm building a midi controller with linear potentiometers, buttons, and LEDs.  I'm using the CD74HC4067M96 analog multiplexer for the potentiometers and was able to really quickly write some code using Control Surface to control Logic Pro with the sliders.  Thanks for making that so easy!  I am also using a 74HC165D 8-bit parallel shift-in register to read the states of the buttons and I don't see a corresponding class for it in the library.  I'm also looking to implement some custom logic for my buttons, meaning I don't want the buttons to simply send 0 or 127 when pressed.  I've already got the code to read the button states, but I'm not sure how I can use them with Control surface to generate CC messages (and taken advantage of the debouncing the library implements).  Should I try to implement a ShiftInRegister class in the library (creating a set of extended IO pins for each button), or is there a way to use the existing Button class with my own read value instead of reading from a specific pin?  Thanks!

**Comment by tttapa** [comment]
> I'm also looking to implement some custom logic for my buttons, meaning I don't want the buttons to simply send 0 or 127 when pressed.

The [Custom-MIDI-Output-Element](https://tttapa.github.io/Control-Surface-doc/Doxygen/d4/d3b/Custom-MIDI-Output-Element_8ino-example.html) would be a good place to start.

> I've already got the code to read the button states, but I'm not sure how I can use them with Control surface to generate CC messages

You can send CC messages using [`Control_Surface.sendCC(address, value);`](https://tttapa.github.io/Control-Surface-doc/Doxygen/d6/d72/classMIDI__Sender.html#a82b919f54a86238cc55ae5ed7eeac578) where the address is of type `MIDIAddress` (see [here](https://tttapa.github.io/Control-Surface-doc/Doxygen/d6/dc6/md_pages_Basics.html)), and value is between 0 and 127. 

> (and taken advantage of the debouncing the library implements). Should I try to implement a ShiftInRegister class in the library (creating a set of extended IO pins for each button), or is there a way to use the existing Button class with my own read value instead of reading from a specific pin?

If you want to be able to use the existing Button class (and all classes that make use of the Button class), then the only way is to implement your own ExtendedIOElement. You only have to implement the methods you care about, e.g. only `digitalRead()`. 

In this case, you would implement the `updateBufferedInputs()` method to read the shift register state into a member variable. The `digitalReadBuffered()` method then just reads the right bit from that buffered variable. You don't have to implement `digitalRead()` itself, it'll work automatically when the two previous methods are implemented. You'll also need to provide a `begin()` method to initialize the pin mode of the latch pin, initialize the SPI peripheral (or the data and clock pins if you're going to bit-bang). You can provide empty implementations for the other pure virtual methods: `pinModeBuffered`, `digitalWriteBuffered`, `analogWriteBuffered`, `analogReadBuffered` and `updateBufferedOutputs`. I'm not sure how familiar you are with inheritance and virtual methods in C++, so if you have trouble getting started, just let me know and I'll see if I can point you to the right resources.

Do keep in mind that you'll have to add external pull-up resistors to all input pins of the shift register. This is the main reason why I haven't added support for input shift registers yet, a multiplexer is easier to use in my opinion, since you can use a single internal pull-up for all buttons connected to the mux.

**Reply by reznet** [reply]
Thanks for your notes - I had already designed and manufactured the board, along with the pull up resistors.  I was able to implement a shift-in register and have submitted PR #408.  Tests are missing - if you can help with some compilations issues (Arduino library methods I do not seem to exist when compiling for tests) I'm happy to add tests.
